### PR TITLE
feat(aws-ssm): Support parameters by path

### PR DESCRIPTION
**Pull request by rjmsilveira**
This MR is a replacement of my original approach (https://github.com/external-secrets/kubernetes-external-secrets/pull/238 )

I have refactored the code to do have as little changes as possible but still supporting the possibility of fetching secrets from a path instead of a key.

Also added support to fetch paths recursively if needed but the default is 0.

### CODE REVIEW

**Review by martina-if — COMMENTED** [review]
I don't know the code conventions in this repo so I only skimmed through this PR and added some small suggestions.

Thanks for re-working the PR! I am looking forward to seeing this shipped :) It's going to simplify our setup a lot.

File: lib/backends/kv-backend.js [hunk 1]
```diff
@@ -61,7 +81,18 @@ class KVBackend extends AbstractBackend {
         }
       }
 
-      return { [name]: value }
+      if (singleParameterKey) {
+        // Not path, return as is
+        response = { [name]: value }
+      } else {
+        // Returning dict with path keys and values
+        for (var records in value) {
```
Comment (Flydiverny) [inline lib/backends/kv-backend.js:89]:
```suggestion
        for (let records in value) {
```

File: lib/backends/kv-backend.js [hunk 2]
```diff
@@ -26,9 +27,28 @@ class KVBackend extends AbstractBackend {
...
+
+      let response = {}
+      let plainOrObjValue;
+
+      // Supporting fetching by key or by path
+      // If path is not defined, than key will be for sure because of CRD validation
+      let singleParameterKey = true
+      if (path) {
+        singleParameterKey = false
+      }
```
Comment (Flydiverny) [inline lib/backends/kv-backend.js:42]:
```suggestion
      const singleParameterKey = !!path
```

File: lib/backends/kv-backend.js [hunk 3]
```diff
@@ -26,9 +27,28 @@ class KVBackend extends AbstractBackend {
    * @returns {Promise} Promise object representing secret property values.
    */
   _fetchDataValues ({ data, specOptions }) {
+
```
Comment (Flydiverny) [inline lib/backends/kv-backend.js:30]:
```suggestion
```

File: lib/backends/kv-backend.js [hunk 4]
```diff
@@ -26,9 +27,28 @@ class KVBackend extends AbstractBackend {
...
    */
   _fetchDataValues ({ data, specOptions }) {
+    
     return Promise.all(data.map(async dataItem => {
-      const { name, property = null, key, ...keyOptions } = dataItem
-      const plainOrObjValue = await this._get({ key, keyOptions, specOptions })
+      const { name, property = null, key, path, ...keyOptions } = dataItem
+
+      let response = {}
+      let plainOrObjValue;
```
Comment (Flydiverny) [inline lib/backends/kv-backend.js:35]:
```suggestion
      let plainOrObjValue
```

**Review by Flydiverny — COMMENTED** [review]
Thanks for updating this! I like this new approach better of reusing the same backend type and `data` structure.

Would love to see some tests covering this as well!

File: examples/ssm-example.yaml [hunk 5]
```diff
@@ -1,13 +1,19 @@
...
   backendType: systemManager
   # optional: specify role to assume when retrieving the data
   roleArn: arn:aws:iam::123456789012:role/test-role
   # optional: specify region
-  region: eu-west-1
+  region: us-west-2
   data:
-    - key: /foo/name1
+    # can either be key+name or all keys from a given path or even both
+    # order is important if you have same key name on different paths that you want to include
```
[the remark this answers is not in the document]
Comment (rjmsilveira) [reply examples/ssm-example.yaml]: Thanks for the suggestion. I updated it just now :)

File: examples/ssm-example.yaml [hunk 6]
```diff
@@ -1,13 +1,19 @@
...
+    - key: /foo/name
       name: variable-name
+    - path: /bar/
+      # optional: choose wether to scrape all child paths or not. Default is false
```
[the remark this answers is not in the document]
Comment (rjmsilveira) [reply examples/ssm-example.yaml]: Thanks for the suggestion. Updated :)

File: lib/backends/kv-backend.js [hunk 7]
```diff
@@ -26,9 +27,28 @@ class KVBackend extends AbstractBackend {
...
+
+      // Supporting fetching by key or by path
+      // If path is not defined, than key will be for sure because of CRD validation
```
[the remark this answers is not in the document]
Comment (rjmsilveira) [reply lib/backends/kv-backend.js]: Thanks for the suggestion. Updated :)

File: lib/backends/kv-backend.js [hunk 8]
```diff
@@ -26,9 +27,28 @@ class KVBackend extends AbstractBackend {
...
+      if (path) {
+        singleParameterKey = false
+      }
+
+      if (singleParameterKey) {
+        // Single secret
+        plainOrObjValue = await this._get({ key, keyOptions, specOptions })
+      } else {
+        // Whole path
+        plainOrObjValue = await this._getByPath({ path, keyOptions, specOptions })
```
[the remark this answers is not in the document]
Comment (rjmsilveira) [reply lib/backends/kv-backend.js:45]: Thanks for all the suggestions. As mentioned, I am not a developer and I appreciate these tips.
All were updated.
Regarding this graceful exit, do you have any suggestions on how to achieve that?

[the remark this answers is not in the document]
Comment (rjmsilveira) [reply lib/backends/kv-backend.js:45]: Meanwhile I believe I now understood what you meant :) thanks for the suggestion and it should be fixed now

**Comment by rjmsilveira** [comment]
@Flydiverny created one test for this new feature. Let me know if it's enough :smile:

File: examples/ssm-example.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (martina-iglesias) [reply examples/ssm-example.yaml]: Looks great, thank you!

[the remark this answers is not in the document]
Comment (martina-if) [reply examples/ssm-example.yaml]: Looks great, thank you!

**Comment by komljen** [comment]
@rjmsilveira I'm trying to get us off our fork where we use data by path, but seems like I cannot use this approach as I get the following error `spec.data.name: Required value, spec.data.key: Required value`.

Why would those be required when the only path is used?

**Comment by komljen** [comment]
> @rjmsilveira I'm trying to get us off our fork where we use data by path, but seems like I cannot use this approach as I get the following error `spec.data.name: Required value, spec.data.key: Required value`.
> 
> Why would those be required when the only path is used?

I forgot to update CRD, all good now!

**Pull request merged**